### ISSUE TITLE
🐛 --navigate n/N jumps to next/prev hunk, not file?

### ISSUE THREAD

**Issue by carlpett** [dandavison/delta#1666]
Possibly a misunderstanding? But when I turn on navigation, do `git diff` and press `n`, the screen advances to the next hunk, even if it is in the same file.

I'm using delta 0.16.5, from Fedora repos. Running inside tmux on Alacritty, if that is relevant.

**Comment by jasonmobley** [comment]
Seeing the same in delta 0.17.0

**Comment by sadamczyk** [comment]
I'm a bit annoyed by this as well. Jumping to the next hunk is usually such a small jump that scrolling with Ctrl+D/U already is more than good enough. And it's especially annoying with large files like lock files.

But this behavior is documented in the _Features_ list of the [README.md](https://github.com/dandavison/delta?tab=readme-ov-file#features) (even though it's missing from the [docs](https://dandavison.github.io/delta/navigation-keybindings-for-large-diffs.html)):
> `n` and `N` keybindings to move between files in large diffs, and **between diffs** in `log -p` views (`--navigate`)

There are 2 options to fix this:
1. Set `--hunk-label ""` to an empty string. Or via `~/.gitconfig`: `hunk-label =` (just nothing).
This removes the `•` from the beginning of every hunk, but that's totally fine for me. :shrug: 
2. Set the `--navigate-regex` value to not search for the hunk labels.
`~/.git/config`: `navigate-regex = ^(commit|added:|removed:|renamed:|Δ)`
But this assumes that you didn't change any of the configs that go into `navigate-regex` by default, like `--file-added-label` etc, since `navigate-regex` is actually built from all those options, see https://github.com/dandavison/delta/blob/0f8a2f0786d56e9395b7e7f244f356bfdab562e8/src/features/navigate.rs#L51-L58 
So if you ever change any of these options, you'll need to manually change the `navigate-regex` option to reflect that as well. Otherwise this seems like the cleanest option.

**Comment by dandavison** [comment]
Just want to say that I sympathise with this! I also find that if I'm using navigate to jump, then by-hunk jumps often feel too small. So I'm open to interface design change proposals here.  

A related variant is to increase our diff context, i.e. with `git diff|show -U99 | delta` or

```gitconfig
[diff]
    context = 99
```

If we set that number large enough, then there are no within-file hunks any longer hence navigate jumps become by-file! Of course it can sometimes not be exactly what you want to have to scroll past large sections of unchanged content, but sometimes the lack of hunk breaks makes diffs easier to understand.

**Comment by sadamczyk** [comment]
@dandavison Good idea with setting a wider context. I'll definitely play around with increasing it a bit, even though 99 seems way too high for me personally. I'll set it to 9 and see how I like it, which will still reduce some hunks hopefully.

IMO The best way would be to have multiple search patterns, that you can search via different hotkeys.
For example something like `n` / `N` just for hunk jumps, `c` / `C` for commits, `f` / `F` for files etc.
That way the user can use whatever jumps make the most sense for their current use case.
But since delta uses plain old less, this isn't really possible afaik. 

I _could_ almost emulate this behavior with custom terminal hotkeys (I use WezTerm), so that my terminal inputs the correct search pattern for whatever navigation "mode" I want to use.
So for example pressing Ctrl+Shift+H could send the following key strokes to the terminal and therefore set less' search pattern to "hunks only": `/^(•)↵` And then just use `n`/`N` as always.
And then have other hotkeys for other common navigation/search patterns I use.

**Comment by unphased** [comment]
I don't need the n/N bind for navigate to do anything other than hop between commits, and commits only. I see why hopping by files would be useful (hopping hunks is decidedly not useful), but the really useful shortcut is by commit. 

To do that I just set the delta config: `navigate-regex = "^commit "` 

One more enhancement would be if we can get a few added binds for less so that we can have four binds and not just two, so we can hop to next/prev commit and next/prev file.

**Comment by nnutter** [comment]
@dandavison what if there were just keybindings to go to next file `f` / `F` (and next commit `c` / `C` in my opinion)? `f` / `F` does collide with `less` keybindings but maybe people can just opt into it?

**Comment by dandavison** [comment]
@nnutter delta is just a program that takes stdin and outputs stdout; it does not respond to keybindings. It does however replace itself with a less process, which does have keybindings. `n`/`N` are less keybindings.

In other words, delta is doing all it can AFAIK -- it's desperately making the best use it can of the available functionality bound to keys by less. To do more would require a dedicated TUI. Now that AI can protoype and write code much faster than humans, I am personally experimenting with using [gitu](https://github.com/altsem/gitu/issues/49#issuecomment-5593509861) as the TUI, and having it call delta.
